RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplateReco2D.cc

```diff
@@ -0,0 +1,654 @@
+   const float fracpix = 0.45f;
+   
+   
+   // Extract some relevant info from the 2D template
+   
+   if(id > 0) {templ2D.interpolate(id, cotalpha, cotbeta, locBz, locBx);}
+   float xsize = templ2D.xsize();
+   float ysize = templ2D.ysize();
+   
+   // Allow Qbin Q/Q_avg fractions to vary to optimize error estimation
+   
+   for(i=0; i<3; ++i) {fbin[i] = templ2D.fbin(i);}
+   
+   float q50 = templ2D.s50();
+   float pseudopix = 0.1f*q50;
+   float pseudopix2 = q50*q50;
+   
+   
+   // Get charge scaling factor
+   
+   float qscale = templ2D.qscale();
+   
+   // Check that the cluster container is (up to) a 7x21 matrix and matches the dimensions of the double pixel flags
+   
+   int nclusx = cluster.mrow;
+   int nclusy = (int)cluster.mcol;
+   bool * xdouble = cluster.xdouble;
+   bool * ydouble = cluster.ydouble;
+   
+   // First, rescale all pixel charges and compute total charge
+   float qtotal = 0.f;
+   for(i=0; i<nclusx*nclusy; ++i) {
+      cluster.matrix[i] *= qscale;
+      qtotal +=cluster.matrix[i];
+   }
+   
+   // uncertainty and final corrections depend upon total charge bin
+   
+   float fq0 = qtotal/templ2D.qavg();
+   
+   // Next, copy the cluster and double pixel flags into a shifted workspace to
+   // allow for significant zeros [pseudopixels] to be added
+   
+   float clusxy[BXM2][BYM2];
+   for(j=0; j<BXM2; ++j) {for(i=0; i<BYM2; ++i) {clusxy[j][i] = 0.f;}}
+   
+   int indexxy[2][NPIXMAX];
+   float pixel[NPIXMAX];
+   float sigma2[NPIXMAX];
+   float minmax = templ2D.pixmax();
+   int imin=BYM2, imax=0, jmin=BXM2, jmax=0;
+   float ylow0 = 0.f, yhigh0 = 0.f, xlow0 = 0.f, xhigh0 = 0.f;
+   float ye = 0.f, xe;
+   int npixel = 0;
+   float ysum[BYM2], ypos[BYM2];
+   //   float xq = 0.f, qtot = 0.f;
+   
+   //Truncate pixels and make a y-projection
+   
+   ysum[0] = 0.f;
+   ypos[0] = -0.5f*ysize;
+   for(i=1; i<nclusy+1; ++i) {
+      float ypitch = ysize;
+      float maxpix = minmax;
+      if(ydouble[i-1]) {ypitch += ysize; maxpix *=2.f;}
+      ypos[i] = ye + ypitch/2.;
+      xe = 0.f;
+      ysum[i] = 0.f;
+      for(j=1; j<nclusx+1; ++j) {
+         float xpitch = xsize;
+         if(xdouble[j-1]) {xpitch += xsize;}
+         // Truncate pixel charges
+         if(cluster(j-1,i-1) > maxpix) {
+            clusxy[j][i] = maxpix;
+         } else {
+            clusxy[j][i] = cluster(j-1,i-1);
+         }
+         if(clusxy[j][i] > 0.) {
+            ysum[i] += clusxy[j][i];
+            //          xq += (xe + xpitch/2.)*clusxy[j][i];
+            if(j < jmin) {jmin = j; xlow0 = xe;}
+            if(j > jmax) {jmax = j; xhigh0 = xe+xpitch;}
+            if(i < imin) {imin = i; ylow0 = ye;}
+            if(i > imax) {imax = i; yhigh0 = ye+ypitch;}
+            indexxy[0][npixel] = j;
+            indexxy[1][npixel] = i;
+            pixel[npixel] = clusxy[j][i];
+            ++npixel;
+         }
+         xe += xpitch;
+      }
+      //      qtot += ysum[i];

```

Comment:
Are these commented out lines needed? If so, please add another line of comment which explain why; otherwise just remove them

Reply:
I don't know.  I'll need to discuss with Morris.  I propose to postpone this until later.